server/models/groups/removeMembers.js

```diff
@@ -0,0 +1,67 @@
+const { pool } = require('../database')
+
+/**
+ *
+ * @param {*} param0
+ * @param {Object} param0.params
+ * @param {Object} param0.body
+ * @param {Object} param0.username
+ * @return {Promise}
+ *
+ */
+
+function removeMembers({ params, body, username }) {
+  return new Promise((resolve, reject) => {
+    const { group_id: groupId } = params
+    const { members } = body
+    pool.query(
+      `SELECT is_group_moderator FROM user_groups WHERE username=? AND group_id=?`,
+      [username, groupId],
+      (error, results) => {
+        if (error) {
+          return reject(error)
+        }
+        const { is_group_moderator } = results[0]
+        if (is_group_moderator === 0) {
+          return reject(`The user do not have moderator access of the group!`)
+        }
+        let invalidUsernames = []
+        let processedUsernames = 0
+        members.forEach((member) => {
+          pool.query(
+            `DELETE FROM user_groups WHERE username=? AND group_id=? 
+                AND username NOT IN (SELECT creator FROM \`groups\` WHERE id=?)`,
```

Comment:
Each time running this query seems a bit expensive :
```
SELECT creator FROM \`groups\` WHERE id=?`
```
It would be better if we run this query once, outside the loop and then use its result inside the loop, that would be much faster.
And inside the loop, we can check whether the request member is creator or not, directly by `if` statement, which would be a lot faster than checking it in the SQL query for all the rows.
Inside the loop, only this query would run -
```
DELETE FROM user_groups WHERE username=? AND group_id=? 
```

Reply:
resolved